firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +159,86 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
     return Tasks.call(executor, this::deleteFirebaseInstallationId);
   }
 
-  /** Returns the application id of the {@link FirebaseApp} of this {@link FirebaseInstallations} */
-  @VisibleForTesting
-  String getApplicationId() {
-    return firebaseApp.getOptions().getApplicationId();
-  }
-
-  /** Returns the nick name of the {@link FirebaseApp} of this {@link FirebaseInstallations} */
-  @VisibleForTesting
-  String getName() {
-    return firebaseApp.getName();
-  }
-
-  /**
-   * Returns the {@link PersistedFidEntry} from shared prefs.
-   *
-   * @throws {@link FirebaseInstallationsException} when shared pref is empty or {@link
-   *     PersistedFidEntry} is in error state.
-   */
-  private PersistedFidEntry getPersistedFid() throws FirebaseInstallationsException {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private Task<String> addGetIdListener() {
+    TaskCompletionSource<String> taskCompletionSource = new TaskCompletionSource<>();
+    StateListener l = new GetIdListener(taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
     }
-    return persistedFidEntry;
+    return taskCompletionSource.getTask();
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
+  private Task<InstallationTokenResult> addGetAuthTokenListener() {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+          break;
```

Comment:
I'm not sure whether we should keep the `break` or not. If we don't, listeners will finish earlier than they would normally which seems reasonable to me. WDYT?

Reply:
I dint think of the ```break;``` until my ```testGetAuthToken_multipleCallsForceRefresh_fetchedNewTokenTwice()``` test failed. 

Reason: Multiple getAuthToken(FOREC_REFRESH) calls - if first call is completed it triggers onStateReached for both calls that are in the list . 

GetAuthTokenListener onStateReached condition will pass (Registered & Auth Toke is valid) and remove both listeners from the list. 

However, the second call execution continues, generates new auth token and updates the shared pref with new values - resulting in discrepancy in what we report and we persist in FIS DB & local storage.

Reply:
ignoring what the code actually does atm, what's the expected behavior of calling getAuthToken(FORCE_REFRESH) twice? Are we expecting the first Task to be completed with a soon-to-be invalid auth token and the second task will be completed with the newly-valid token? Or do we want both tasks to be completed with the same auth token?

Reply:
AFIK  we are expected to return the first Task on completion with a soon-to-be invalid auth token and the second task will be completed with the newly-valid token.  I verified this flow with Di and yet to verify it with Rayo.

IMO it would be good to have both tasks be completed with the same auth token. But wont we complicate the logic ( like Force Refresh behaves differently when one call is made vs multiple calls, or when multiple calls with do-not-force-refresh and force-refresh options are called).

Your thoughts.

Reply:
If that's the expected and desired outcome then I guess this is fine. I think it's a little error prone, but that's a product decision for your team and not a technical one that should be resolved here.

I think it would be clearer if we completed both tasks with the same result where possible, but to do that we'd have to keep the current auth token in memory and read it as part of our listener creation routine. I don't think that's prohibitively complicated (and it means we could simplify elsewhere by removing the break here and the conditional below), but that's a decision for your team.

Reply:
This is one corner case. I am yet to discuss and decide the outcome with the team. 

However, I tried incorporating your suggestion. Can you PTAL if the changes are correct now.